viewing_party/party.py

```diff
@@ -0,0 +1,243 @@
+
+
+
+
+
+################ WAVE 2 ##############
+
+
+#wave 2, part 1
+def get_watched_avg_rating(user_data):
+    """Calculates avd movie rating and returns."""
+
+    rating_sum = 0.0
+    rating_avg = 0.0
+    
+    if len(user_data["watched"]) == 0:
+        return rating_avg
+
+    for movie_dict in user_data["watched"]:
+        rating_sum += movie_dict["rating"]
+
+    rating_avg = rating_sum / (len(user_data["watched"]))
+    return rating_avg
+
+
+#wave 2, part 2
+def get_most_watched_genre(user_data):
+    """Decides most popular genre from watched list. Returns string."""
+
+    genre_list = []
+    fantasy_count = 0
+    action_count = 0
+    intrigue_count = 0
+
+    if len(user_data["watched"]) == 0:
+        return None
+
+    for movie_dict in user_data["watched"]:
+        genre_list.append(movie_dict["genre"])
+
+    fantasy_count = genre_list.count("Fantasy")
+    action_count = genre_list.count("Action")
+    intrigue_count = genre_list.count("Intrigue")
+
+    if (fantasy_count > action_count) and (fantasy_count > intrigue_count):
+        return "Fantasy"
+    elif (action_count > fantasy_count) and (action_count > intrigue_count):
+        return "Action"
+    else: 
+        return "Intrigue"
+
+
+
+
+
+
+
+
+################ WAVE 3 ##############
+
+#wave 3, part 1
+def get_unique_watched(user_data):
+    """Takes dict, returns list of dicts that only user has seen."""
+    
+    only_user_watched = []
+    user_watched_set = set()
+    friends_watched_set = set()
+    only_user_watched_set = set()
+
+    for movie_user_watched in user_data["watched"]:
+        user_watched_set.add(movie_user_watched["title"])
+        for friends_watched_lists in user_data["friends"]:
+            for movies_friends_watched_list in friends_watched_lists.values():
+                for indiv_dict in movies_friends_watched_list:
+                    friends_watched_set.add(indiv_dict["title"])
+
+    only_user_watched_set = user_watched_set.difference(friends_watched_set)
+
+    for item in only_user_watched_set:
+        only_user_watched.append({"title" : item})
+
+    return only_user_watched
+                    
+    
+
+#wave 3, part 2
+def get_friends_unique_watched(user_data):
+    """Takes dict, returns list of dicts that only friends have seen."""
+
+    user_watched_set = set()
+    friends_watched_set = set()
+    only_friends_watched_set = set()
+    only_friends_watched = []
+
+    for movie_user_watched in user_data["watched"]:
+        user_watched_set.add(movie_user_watched["title"])
+        
+    for friends_watched_lists in user_data["friends"]:
+        for movies_friends_watched_list in friends_watched_lists.values():
+            for indiv_dict in movies_friends_watched_list:
+                friends_watched_set.add(indiv_dict["title"])
+
+    only_friends_watched_set = friends_watched_set.difference(user_watched_set)
+
+    for item in only_friends_watched_set:
+        only_friends_watched.append({"title" : item})
```

Comment:
The main difference between my previous feedback and here is that if we want to make sure the friend output list is unique. We would need to track the friend titles we've added to the output list, something like this:

```python
    unique_output_titles = set()
    for friends_watched_lists in user_data["friends"]:
        for indiv_dict in friends_watched_lists["watched"]:
            if indiv_dict["title"] not in unique_output_titles and indiv_dict["title"] not in user_watched_set:
                only_friends_watched.append(indiv_dict["title"])
                unique_output_titles.add(indiv_dict["title"])

```